maven/JarAssembler.kt

```diff
@@ -34,35 +36,73 @@ class JarAssembler : Callable<Unit> {
+
+    /**
+     * For path "a/b/c.java" inserts "a/" and "a/b/ into `entries`
+     */
+    private fun preCreateDirectories(path: Path) {
+        for (i in path.nameCount-1 downTo 1) {
+            val subPath = path.subpath(0, i).toString() + "/"
+            entries[subPath] = ByteArray(0)
+        }
+    }
+
+    private fun getFinalPath(entry: ZipEntry, sourceFileBytes: ByteArray): String {
+        return if (entry.name.endsWith(".java")) {
+            // files in source JARs are moved according to their `package` statement
+            val sourceFile = sourceFileBytes.toString(Charset.forName("UTF-8"))
+            val sourceFileName = Paths.get(entry.name).fileName.toString()
+            val sourceFilePackage = javaPackageRegex.find(sourceFile)?.groups?.get(1)?.value ?: throw RuntimeException("could not obtain package of ${entry.name}")
+            "${sourceFilePackage.replace(".", "/")}/$sourceFileName"
+        } else {
+            entry.name
+        }
+    }
+
+    private fun prepareEntry(inputStream: BufferedInputStream, entry: ZipEntry) {
+        val sourceFileBytes = inputStream.readBytes()
+        val resultLocation = getFinalPath(entry, sourceFileBytes)
+        entries[resultLocation] = sourceFileBytes
+        preCreateDirectories(Paths.get(resultLocation))
+    }
 
     override fun call() {
         ZipOutputStream(BufferedOutputStream(FileOutputStream(output_file))).use { out ->
             if (pomFile != null) {
-                val pomFileEntry = ZipEntry("META-INF/maven/${groupId}/${artifactId}/pom.xml")
-                out.putNextEntry(pomFileEntry)
-                out.write(pomFile!!.readBytes())
+                val pomPath = "META-INF/maven/${groupId}/${artifactId}/pom.xml"
+                entries[pomPath] = pomFile!!.readBytes()
+                preCreateDirectories(Paths.get(pomPath))
             }
             for (jar in jars) {
                 ZipFile(jar).use { jarZip ->
                     jarZip.entries().asSequence().forEach { entry ->
                         if (entry.name.contains("META-INF")) {
+                            // pom.xml will be added by us
                             return@forEach
                         }
                         if (entryNames.contains(entry.name)) {
+                            throw RuntimeException("duplicate entry in the JAR: ${entry.name}")
+                        }
+                        if (entry.isDirectory) {
+                            // needed directories would be added by us
                             return@forEach
                         }
                         entryNames.add(entry.name)
                         BufferedInputStream(jarZip.getInputStream(entry)).use { inputStream ->
-                            val newEntry = ZipEntry(prefix + entry.name)
-                            out.putNextEntry(newEntry)
-                            inputStream.copyTo(out, 1024)
+                            prepareEntry(inputStream, entry)
```

Comment:
I think the name `prepareEntry` isn't very meaningful and too generic. If I see a step "prepare", I would want to see another step "execute" and "cleanup", for example. That way, I understand that the flow of the program follows the three-stage approach of "prepare", "execute" and "cleanup".

In our program here, we only have "prepare". Therefore it actually makes the reader wondering.

Therefore I would suggest inlining the content here:

```
BufferedInputSttream(...).use { inputStream ->
        val sourceFileBytes = inputStream.readBytes()
        val resultLocation = getFinalPath(entry, sourceFileBytes)
        entries[resultLocation] = sourceFileBytes
        preCreateDirectories(Paths.get(resultLocation))
}
```

which is already clean enough since it only has 4 lines

Reply:
@lolski addressed in https://github.com/vaticle/bazel-distribution/pull/305/commits/cad87304cd7785647fcfc952b1fd3152cf396490